scipy/linalg/_matfuncs_sqrtm.h

```diff
@@ -0,0 +1,186 @@
+*/
+static PyObject*
+recursive_schur_sqrtm(PyObject *dummy, PyObject *args) {
+    int isComplex = 0, isIllconditioned = 0, isSingular = 0, info = 0;
+    PyArrayObject *ap_Am=NULL;
+    void* mem_ret = NULL;
+    // Get the input array
+    if (!PyArg_ParseTuple(args, ("O!"), &PyArray_Type, (PyObject **)&ap_Am))
+    {
+        return NULL;
+    }
+
+    // Check for dtype compatibility
+    if ((PyArray_TYPE(ap_Am) != NPY_FLOAT64) &&
+         (PyArray_TYPE(ap_Am) != NPY_FLOAT32) &&
+         (PyArray_TYPE(ap_Am) != NPY_COMPLEX64) &&
+         (PyArray_TYPE(ap_Am) != NPY_COMPLEX128)
+       )
+    {
+        PYERR(sqrtm_error, "Input must be nD array of type "
+              "float32, float64, complex64, or complex128.");
+    }
+
+    int ndim = PyArray_NDIM(ap_Am);              // Number of dimensions
+    npy_intp* shape = PyArray_SHAPE(ap_Am);      // Array shape
+    npy_intp n = shape[ndim - 1];                // Slice size
+    int input_type = PyArray_TYPE(ap_Am);        // Data type enum value
+    // Compare last two dimensions for squareness
+    if (n != shape[ndim - 2])
+    {
+        PYERR(sqrtm_error, "Last two dimensions of the input must be the same.");
+    }
+
+    // Create the output array with the same shape as the input with twice the
+    // number of entries to accomodate for potential complex-valued data from
+    // real data which will be cast to complex e.g., "ret.asview(complex128)"
+    // Example, (3, 3) -> (18), (4, 5, 5) -> (4, 50)
+    npy_intp ret_dims = 1;
+    for (int i = 0; i < ndim; i++) { ret_dims *= shape[i]; }
+    if ((input_type == NPY_FLOAT32) || (input_type == NPY_FLOAT64))
+    {
+        ret_dims *= 2;
+    }
+
+    if (PyArray_TYPE(ap_Am) == NPY_FLOAT32)
+    {
+        mem_ret = malloc(ret_dims*sizeof(float));
+        matrix_squareroot_s(ap_Am, (float*)mem_ret, &isIllconditioned, &isSingular, &info, &isComplex);
+
+    } else if (PyArray_TYPE(ap_Am) == NPY_FLOAT64) {
+
+        mem_ret = malloc(ret_dims*sizeof(double));
+        matrix_squareroot_d(ap_Am, (double*)mem_ret, &isIllconditioned, &isSingular, &info, &isComplex);
+
+    } else if (PyArray_TYPE(ap_Am) == NPY_COMPLEX64) {
+
+        mem_ret = malloc(ret_dims*sizeof(SCIPY_C));
+        matrix_squareroot_c(ap_Am, (SCIPY_C*)mem_ret, &isIllconditioned, &isSingular, &info, &isComplex);
+        isComplex = 1;
+
+    } else if (PyArray_TYPE(ap_Am) == NPY_COMPLEX128) {
+
+        mem_ret = malloc(ret_dims*sizeof(SCIPY_Z));
+        matrix_squareroot_z(ap_Am, (SCIPY_Z*)mem_ret, &isIllconditioned, &isSingular, &info, &isComplex);
+        isComplex = 1;
+
+    } else {
+        PYERR(sqrtm_error, "Unsupported input data type to sqrtm C function.");
+    }
+
+    if (info < 0)
+    {
+        // Internal failure memory or LAPACK error, fail and return the error code in info
+        free(mem_ret);
+        Py_INCREF(Py_None);
+        return Py_BuildValue("Niii", Py_None, isIllconditioned, isSingular, info);
+    }
```

Comment:
Please consider either raising an informative exception or returning an array full of nans on failure.

Reply:
This is something that came sometime ago but I can't find the relevant issue. If it is singular the results are not correct but it could be correct for a part of the array. I checked matlab/octave and other places and they don't fill in the array rather let the partially correct parts stay intact and return whatever has happened internally.

Reply:
Whoa. So you're talking about a situation where in a batch of size 25 (`sqrtm(a)` with `a.shape = (25, n, n)`), 24 elements are OK but the 25-th is bad, and you want to make these 24 computed elements available to the user?

Would be nice to establish an invariant from the get-go (for otherwise we'll never get to it). And document it, too.

Reply:
Yes not only that but also within the same slice like so in matlab

![image](https://github.com/user-attachments/assets/054022b0-6615-41b8-9cbd-cb7db6921e6f)